Script for training VISH

saadhamid96 opened the pull request.

saadhamid96 committed:
Add script that should run VISH. There is still buggy.

saadhamid96 committed:
Fix bug by passing compile=False flags into optimisation routine.

saadhamid96 committed:
Separate VISH components into their own module, and add warped version

saadhamid96 committed:
Fix map_to_sphere function

vdutor commented on gspheres/vish.py:
```diff
@@ -0,0 +1,229 @@
+
+
+def gauss_kl_vish(q_mu, q_sqrt, K):
+    """
```
```suggestion
    """
    Author: ST--
    Source: GPflow notebook on variational Fourier features.
```

vdutor commented on gspheres/vish.py:
```diff
@@ -0,0 +1,229 @@
+        jitter=None
+):
+    """Covariance matrix between spherical harmonic features."""
+    eigenvalues = kernel.eigenvalues(len(inducing_variable))
```
This doesn't look exactly right. `kernel.eigenvalues(n)` returns the eigenvalues `\lambda_n`'s for the first n levels. However, for each level, we have `N(d, n)` number of spherical harmonics. We need to make sure to have the correct repetition of eigenvalues. I've done this before using the "chain" function: https://github.com/vdutor/GPsOnSpheres/blob/master/scripts/plot_samples.py#L14.

saadhamid96 replied:
```suggestion
    eigenvalues_per_level = kernel.eigenvalues(inducing_variable.max_degree)
    num_harmonics_per_level = tf.convert_to_tensor(
        [num_harmonics(dimension, n) for n in range(inducing_variable.max_degree)]
    )
    eigenvalues = chain(eigenvalues_per_level, num_harmonics_per_level)
```

Okay, cheers for pointing that out, I think that your code from plot_samples can just be copied over, and we'll add max_degree as an attribute of inducing_variable upon instantiation.

vdutor replied:
Nice, feel free to directly commit changes to this branch.

vdutor commented on gspheres/vish.py:
```diff
@@ -0,0 +1,229 @@
+    Covariance between spherical harmonic features and function values.
+
+    """
+    return tf.transpose(inducing_variable.spherical_harmonics(X))
```
which return shape are we expecting here?

saadhamid96 replied:
Number of inducing variables (rows) by number of points in X (columns).

vdutor commented on gspheres/vish.py:
```diff
@@ -0,0 +1,229 @@
+        super().__init__(*args, **kwargs)
+        self.alpha = alpha
+
+    def predict_f_sq(
```
this `elbo` computation will still use the default`predict_f` (without the rescaling) so I think we have to do the rescaling in the `conditional` instead.

saadhamid96 replied:
Yes, that is what I was intending. We warp the data, fit the SVGP to the warped data as normal, and then transform the posterior.

vdutor replied:
What exactly do you mean by warp the data? Do you mean normalising? Because as far as I can see Y does not get warped in the script `scripts/vish.py`.

vdutor reviewed (commented):
Hey Saad, this is looking really good! I think the only conceptual mistake I spotted is in the computation of the `Kuu`, where we need to account for the repetition of eigenvalues within each level. I would also suggest creating a module `vish` with a proper `__init__.py` and split up the large `vish.py` file in a `conditional.py` and `util.py` file (for example). This will probably help us down the line if we need to add more logic.

saadhamid96 committed:
Update docsting for gauss_kl_vish in gspheres/vish.py

Co-authored-by: Vincent Dutordoir <dutordoirv@gmail.com>

saadhamid96 commented on gspheres/vish.py:
```diff
@@ -0,0 +1,230 @@
+    """Wraps SphericalHarmonics."""
+    def __init__(self, dimension, degrees):
+        self.spherical_harmonics = SphericalHarmonics(dimension, degrees)
+
```
```suggestion
        self.max_degree = degrees

```

vdutor reviewed (commented):
Looking good. Can I also recommend adding some simple unit tests (see tests directory). A test I often add is comparing VISH with its diagonal Kuu to another VISH model where we use a full Kuu matrix where only the diagonal elements are non-zero. The full Kuu VISH can use the plain conditional and KL from GPflow but the ELBO should be identical. These tests will probably help us a lot in the future when making modifications to the code to make sure we haven't changed functionality.

If you want to visually inspect the model and see if it makes sense, I would recommend modelling a toy dataset on the sphere first. I added quite a bit of cool plotting functionality for this (see scripts/plotting*). 

Please let me know how I can best help.

saadhamid96 committed:
Correct repetition of eigenvalues in Kuu matrix

saadhamid96 commented:
Okay, adding some tests and splitting things into modules is a good idea. I started a new branch and added mixed eigenfunction and pseudo-observation inducing variables yesterday. It seems to be functional, though optimisation of the inducing variables is hard. I'll open a new pull request from that branch, and we can do the adding of the tests there?

On the warping: scripts/vish.py was just to check everything ran properly. I have been working on a script to compare different models in the vish-pi repo, and that is a good example. [https://github.com/saadhamidml/vish-pi/blob/main/src/vish_local_evaluation.py](url). Note that the latest version of the vish_local_evaluation.py script requires the latest version of the mixed_features branch in GPsOnSpheres. I've also added a sub-page on the Notion to talk about the mixed features [https://www.notion.so/bayesxl/Mixed-Eigenfunction-and-Pseudo-Observation-Inducing-Variables-02ef5ee4c32a4ae69927ad70489e1416](url). If you have any comments on something I may have been getting wrong in terms of optimising the mixed inducing variables then please add them! If not, then I think it makes sense to move forward with just eigenfunction inducing variables, since they work pretty well anyway.

If you've got some spare time would you be okay implementing the SphericalMatern for arbitrary dimension? I'll get the acquisition function and each of the problem settings implemented. I'll also start using the GitFlow workflow in the vish-pi repo so that you get notified when I make changes.
